README.md
# hd-key

Bitcoin [BIP32](https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki) hierarchical deterministic keys Typescript implemention.

## Installing / Getting started

```shell
npm install --save hd-key
```

## Basic Usage

### HDKey.fromMasterSeed(seedBuffer[, masterSecret, versions])

```typescript
const seed =
  'a0c42a9c3ac6abf2ba6a9946ae83af18f51bf1c9fa7dacc4c92513cc4dd015834341c775dcd4c0fac73547c5662d81a9e9361a0aac604a73a321bd9103bce8af'
const hdkey = HDKey.fromMasterSeed(Buffer.from(seed, 'hex'))
```

### HDKey.fromExtendedKey(extendedKey[, versions])

```typescript
const key =
  'xprvA2nrNbFZABcdryreWet9Ea4LvTJcGsqrMzxHx98MMrotbir7yrKCEXw7nadnHM8Dq38EGfSh6dqA9QWTyefMLEcBYJUuekgW4BYPJcr9E7j'
const hdkey = HDKey.fromExtendedKey(key)
```

**or**

```typescript
const key =
  'xpub6FnCn6nSzZAw5Tw7cgR9bi15UV96gLZhjDstkXXxvCLsUXBGXPdSnLFbdpq8p9HmGsApME5hQTZ3emM2rnY5agb9rXpVGyy3bdW6EEgAtqt'
const hdkey = HDKey.fromExtendedKey(key)
```

### HDKey.fromJSON(obj)

Creates an `hdkey` object from an object created via `hdkey.toJSON()`.

### hdkey.derive(path)

Derives the `hdkey` at `path` from the current `hdkey`.

```typescript
const seed = 'fffcf9f6f3f0edeae7e4e1dedbd8d5d2cfccc9c6c3c0bdbab7b4b1aeaba8a5a29f9c999693908d8a8784817e7b7875726f6c696663605d5a5754514e4b484542'
const hdkey = HDKey.fromMasterSeed(Buffer.from(seed, 'hex'))
const childkey = hdkey.derive("m/0/2147483647'/1")

console.log(childkey.privateExtendedKey)
// -> "xprv9zFnWC6h2cLgpmSA46vutJzBcfJ8yaJGg8cX1e5StJh45BBciYTRXSd25UEPVuesF9yog62tGAQtHjXajPPdbRCHuWS6T8XA2ECKADdw4Ef"
console.log(childkey.publicExtendedKey)
// -> "xpub6DF8uhdarytz3FWdA8TvFSvvAh8dP3283MY7p2V4SeE2wyWmG5mg5EwVvmdMVCQcoNJxGoWaU9DCWh89LojfZ537wTfunKau47EL2dhHKon"
```

### hdkey.sign(hash)

Signs the buffer `hash` with the private key using `secp256k1` and returns the signature as a buffer.

### hdkey.verify(hash, signature)

Verifies that the `signature` is valid for `hash` and the `hdkey`'s public key using `secp256k1`. Returns `true` for valid, `false` for invalid. Throws if the `hash` or `signature` is the wrong length.

### hdkey.wipePrivateData()

Wipes all record of the private key from the `hdkey` instance. After calling this method, the instance will behave as if it was created via `HDKey.fromExtendedKey(xpub)`.

### hdkey.toJSON()

Serializes the `hdkey` to an object that can be `JSON.stringify()`ed.

```typescript
const seed = 'fffcf9f6f3f0edeae7e4e1dedbd8d5d2cfccc9c6c3c0bdbab7b4b1aeaba8a5a29f9c999693908d8a8784817e7b7875726f6c696663605d5a5754514e4b484542'
const hdkey = HDKey.fromMasterSeed(Buffer.from(seed, 'hex'))

console.log(hdkey.toJSON())
// -> {
//      xpriv: 'xprv9s21ZrQH143K31xYSDQpPDxsXRTUcvj2iNHm5NUtrGiGG5e2DtALGdso3pGz6ssrdK4PFmM8NSpSBHNqPqm55Qn3LqFtT2emdEXVYsCzC2U',
//      xpub: 'xpub661MyMwAqRbcFW31YEwpkMuc5THy2PSt5bDMsktWQcFF8syAmRUapSCGu8ED9W6oDMSgv6Zz8idoc4a6mr8BDzTJY47LJhkJ8UB7WEGuduB'
//    }
```

### hdkey.privateKey

Getter/Setter of the `hdkey`'s private key, stored as a buffer.

### hdkey.publicKey

Getter/Setter of the `hdkey`'s public key, stored as a buffer.

### hdkey.privateExtendedKey

Getter/Setter of the `hdkey`'s `xprv`, stored as a string.

### hdkey.publicExtendedKey

Getter/Setter of the `hdkey`'s `xpub`, stored as a string.

## Developing

### Dev

```shell
yarn watch
```

### Test

```shell
yarn test
```

### Cov

```shel
yarn cov
```

## Versioning

We can maybe use [SemVer](http://semver.org/) for versioning. For the versions available, see the [link to tags on this repository](/tags).

## References

- https://github.com/cryptocoinjs/hdkey

## Licensing

State what the license is and how to find the text version of the license.


CHANGELOG.md
# 0.1.0 (2018-12-25)


### Features

* 🎸 finish ([87b3925](https://github.com/michaelcai/hd-key/commit/87b3925))





config/test.js
'use strict';

// Do this as the first thing so that any code reading it knows the right env.
process.env.BABEL_ENV = 'test';
process.env.NODE_ENV = 'test';
process.env.PUBLIC_URL = '';

// Makes the script crash on unhandled rejections instead of silently
// ignoring them. In the future, promise rejections that are not handled will
// terminate the Node.js process with a non-zero exit code.
process.on('unhandledRejection', err => {
  throw err;
});

const jest = require('jest');
let argv = process.argv.slice(2);

// Watch unless on CI, in coverage mode, or explicitly running all tests
if (
  !process.env.CI &&
  argv.indexOf('--coverage') === -1 &&
  argv.indexOf('--watchAll') === -1
) {
  argv.push('--watch');
}


jest.run(argv);


src/__test__/hdkey.spec.ts
import BigInteger from 'bigi'
import crypto from 'crypto'
import ecurve from 'ecurve'
import HDKey from '../index'
import keys from './keys'

const curve = ecurve.getCurveByName('secp256k1')

describe('hdkey', () => {
  describe('fromMasterSeed', () => {
    keys.valid.forEach(key => {
      it(`should properly derive the chain path: ${key.path}`, () => {
        const hdkey = HDKey.fromMasterSeed(Buffer.from(key.seed, 'hex'))
        const childkey = hdkey.derive(key.path)
        expect(childkey.privateExtendedKey).toEqual(key.private)
        expect(childkey.publicExtendedKey).toEqual(key.public)
      })

      // it("should properly derive the chain path: m/44'/709394", () => {
      //   const hdkey = HDKey.fromMasterSeed(
      //     Buffer.from(
      //       'b851e3ee9054c2d7ad40624f04aa58bea364e7654cc0dbe5a8edcf5315d8f72c6788736e5eb1146cbd8fd4f9604dfb2cf721ac4bbe3ec47ce2a1bf4a8df73195',
      //       'hex'
      //     )
      //   )

      //   const childkey = hdkey.derive("m/44'/709394'/0'/0/1")
      //   expect(childkey.privateKey)

      // })

      describe(`> ${key.path} toJSON() / fromJSON()`, () => {
        it('should return an object read for JSON serialization', () => {
          const hdkey = HDKey.fromMasterSeed(Buffer.from(key.seed, 'hex'))
          const childkey = hdkey.derive(key.path)

          const obj = {
            xpriv: key.private,
            xpub: key.public
          }

          expect(childkey.toJSON()).toEqual(obj)

          const newKey = HDKey.fromJSON(obj)
          expect(newKey.privateExtendedKey).toEqual(key.private)
          expect(newKey.publicExtendedKey).toEqual(key.public)
        })
      })
    })
  })

  describe('- privateKey', () => {
    it('should throw an error if incorrect key size', () => {
      const hdkey = new HDKey()
      expect(() => {
        hdkey.privateKey = Buffer.from([1, 2, 3, 4])
      }).toThrow(/key must be 32/)
    })
  })

  describe('- publicKey', () => {
    it('should throw an error if incorrect key size', () => {
      expect(() => {
        const hdkey = new HDKey()
        hdkey.publicKey = Buffer.from([1, 2, 3, 4])
      }).toThrow(/key must be 33 or 65/)
    })

    it('should not throw if key is 33 bytes (compressed)', () => {
      const priv = crypto.randomBytes(32)
      const pub = curve.G.multiply(BigInteger.fromBuffer(priv)).getEncoded(true)
      expect(pub).toHaveLength(33)
      const hdkey = new HDKey()
      hdkey.publicKey = pub
    })

    it('should not throw if key is 65 bytes (not compressed)', () => {
      const priv = crypto.randomBytes(32)
      const pub = curve.G.multiply(BigInteger.fromBuffer(priv)).getEncoded(
        false
      )
      expect(pub).toHaveLength(65)
      const hdkey = new HDKey()
      hdkey.publicKey = pub
    })
  })

  describe('+ fromExtendedKey()', () => {
    describe('> when private', () => {
      it('should parse it', () => {
        // m/0/2147483647'/1/2147483646'/2
        const key =
          'xprvA2nrNbFZABcdryreWet9Ea4LvTJcGsqrMzxHx98MMrotbir7yrKCEXw7nadnHM8Dq38EGfSh6dqA9QWTyefMLEcBYJUuekgW4BYPJcr9E7j'
        const hdkey = HDKey.fromExtendedKey(key)
        expect(hdkey.versions.private).toEqual(0x0488ade4)
        expect(hdkey.versions.public).toEqual(0x0488b21e)
        expect(hdkey.depth).toEqual(5)
        expect(hdkey.parentFingerprint).toEqual(0x31a507b8)
        expect(hdkey.index).toEqual(2)
        expect(hdkey.chainCode.toString('hex')).toEqual(
          '9452b549be8cea3ecb7a84bec10dcfd94afe4d129ebfd3b3cb58eedf394ed271'
        )
        expect(hdkey.privateKey.toString('hex')).toEqual(
          'bb7d39bdb83ecf58f2fd82b6d918341cbef428661ef01ab97c28a4842125ac23'
        )
        expect(hdkey.publicKey.toString('hex')).toEqual(
          '024d902e1a2fc7a8755ab5b694c575fce742c48d9ff192e63df5193e4c7afe1f9c'
        )
        expect(hdkey.identifier.toString('hex')).toEqual(
          '26132fdbe7bf89cbc64cf8dafa3f9f88b8666220'
        )
      })
    })

    describe('> when public', () => {
      it('should parse it', () => {
        // m/0/2147483647'/1/2147483646'/2
        const key =
          'xpub6FnCn6nSzZAw5Tw7cgR9bi15UV96gLZhjDstkXXxvCLsUXBGXPdSnLFbdpq8p9HmGsApME5hQTZ3emM2rnY5agb9rXpVGyy3bdW6EEgAtqt'
        const hdkey = HDKey.fromExtendedKey(key)
        expect(hdkey.versions.private).toEqual(0x0488ade4)
        expect(hdkey.versions.public).toEqual(0x0488b21e)
        expect(hdkey.depth).toEqual(5)
        expect(hdkey.parentFingerprint).toEqual(0x31a507b8)
        expect(hdkey.index).toEqual(2)
        expect(hdkey.chainCode.toString('hex')).toEqual(
          '9452b549be8cea3ecb7a84bec10dcfd94afe4d129ebfd3b3cb58eedf394ed271'
        )
        expect(hdkey.privateKey).toEqual(null)
        expect(hdkey.publicKey.toString('hex')).toEqual(
          '024d902e1a2fc7a8755ab5b694c575fce742c48d9ff192e63df5193e4c7afe1f9c'
        )
        expect(hdkey.fingerprint).toEqual(638791643)
        expect(hdkey.identifier.toString('hex')).toEqual(
          '26132fdbe7bf89cbc64cf8dafa3f9f88b8666220'
        )
        expect(hdkey.pubKeyHash.toString('hex')).toEqual(
          '26132fdbe7bf89cbc64cf8dafa3f9f88b8666220'
        )
      })
    })
    describe('> when signing', () => {
      it('should work', () => {
        const key =
          'xprvA2nrNbFZABcdryreWet9Ea4LvTJcGsqrMzxHx98MMrotbir7yrKCEXw7nadnHM8Dq38EGfSh6dqA9QWTyefMLEcBYJUuekgW4BYPJcr9E7j'
        const hdkey = HDKey.fromExtendedKey(key)

        const ma = Buffer.alloc(32, 0)
        const mb = Buffer.alloc(32, 8)
        const a = hdkey.sign(ma)
        const b = hdkey.sign(mb)
        expect(a.toString('hex')).toEqual(
          '6ba4e554457ce5c1f1d7dbd10459465e39219eb9084ee23270688cbe0d49b52b7905d5beb28492be439a3250e9359e0390f844321b65f1a88ce07960dd85da06'
        )

        expect(b.toString('hex')).toEqual(
          'dfae85d39b73c9d143403ce472f7c4c8a5032c13d9546030044050e7d39355e47a532e5c0ae2a25392d97f5e55ab1288ef1e08d5c034bad3b0956fbbab73b381'
        )

        expect(hdkey.verify(ma, a)).toBeTruthy()
        expect(hdkey.verify(mb, b)).toBeTruthy()
        expect(
          hdkey.verify(Buffer.alloc(32), Buffer.alloc(64))
        ).not.toBeTruthy()
        expect(hdkey.verify(ma, b)).not.toBeTruthy()
        expect(hdkey.verify(mb, a)).not.toBeTruthy()

        expect(() => {
          hdkey.verify(Buffer.alloc(99), a)
        }).toThrow(/message length is invalid/)

        expect(() => {
          hdkey.verify(ma, Buffer.alloc(99))
        }).toThrow(/signature length is invalid/)
      })
    })

    describe('> when deriving public key', () => {
      it('should work', () => {
        const key =
          'xpub661MyMwAqRbcFtXgS5sYJABqqG9YLmC4Q1Rdap9gSE8NqtwybGhePY2gZ29ESFjqJoCu1Rupje8YtGqsefD265TMg7usUDFdp6W1EGMcet8'
        const hdkey = HDKey.fromExtendedKey(key)

        const path = 'm/3353535/2223/0/99424/4/33'
        const derivedHDKey = hdkey.derive(path)

        const expected =
          'xpub6JdKdVJtdx6sC3nh87pDvnGhotXuU5Kz6Qy7Piy84vUAwWSYShsUGULE8u6gCivTHgz7cCKJHiXaaMeieB4YnoFVAsNgHHKXJ2mN6jCMbH1'
        expect(derivedHDKey.publicExtendedKey).toEqual(expected)
      })
    })

    describe('> when private key integer is less than 32 bytes', () => {
      it('should work', () => {
        const seed = '000102030405060708090a0b0c0d0e0f'
        const masterKey = HDKey.fromMasterSeed(Buffer.from(seed, 'hex'))

        const newKey = masterKey.derive("m/44'/6'/4'")
        const expected =
          'xprv9ymoag6W7cR6KBcJzhCM6qqTrb3rRVVwXKzwNqp1tDWcwierEv3BA9if3ARHMhMPh9u2jNoutcgpUBLMfq3kADDo7LzfoCnhhXMRGX3PXDx'
        expect(newKey.privateExtendedKey).toEqual(expected)
      })
    })

    describe('HARDENED_OFFSET', () => {
      it('should be set', () => {
        expect(HDKey.HARDENED_OFFSET).toBeTruthy()
      })
    })

    describe('> when private key has leading zeros', () => {
      it('will include leading zeros when hashing to derive child', () => {
        const key =
          'xprv9s21ZrQH143K3ckY9DgU79uMTJkQRLdbCCVDh81SnxTgPzLLGax6uHeBULTtaEtcAvKjXfT7ZWtHzKjTpujMkUd9dDb8msDeAfnJxrgAYhr'
        const hdkey = HDKey.fromExtendedKey(key)
        expect(hdkey.privateKey.toString('hex')).toEqual(
          '00000055378cf5fafb56c711c674143f9b0ee82ab0ba2924f19b64f5ae7cdbfd'
        )
        const derived = hdkey.derive("m/44'/0'/0'/0/0'")
        expect(derived.privateKey.toString('hex')).toEqual(
          '3348069561d2a0fb925e74bf198762acc47dce7db27372257d2d959a9e6f8aeb'
        )
      })
    })

    describe('> when private key is null', () => {
      it('privateExtendedKey should return null and not throw', () => {
        const seed = '000102030405060708090a0b0c0d0e0f'
        const masterKey = HDKey.fromMasterSeed(Buffer.from(seed, 'hex'))

        expect(masterKey.privateExtendedKey).toBeTruthy()

        masterKey.wipePrivateData()

        expect(masterKey.privateExtendedKey).not.toBeTruthy()
      })
    })

    describe(' - when the path given to derive contains only the master extended key', () => {
      const hdKeyInstance = HDKey.fromMasterSeed(
        Buffer.from(keys.valid[0].seed, 'hex')
      )

      it('should return the same hdkey instance', () => {
        expect(hdKeyInstance.derive('m')).toEqual(hdKeyInstance)
        expect(hdKeyInstance.derive('M')).toEqual(hdKeyInstance)
        expect(hdKeyInstance.derive("m'")).toEqual(hdKeyInstance)
        expect(hdKeyInstance.derive("M'")).toEqual(hdKeyInstance)
      })
    })

    describe(' - when the path given to derive does not begin with master extended key', () => {
      it('should throw an error', () => {
        expect(() => {
          const hdKey = HDKey.fromMasterSeed(
            Buffer.from(keys.valid[0].seed, 'hex')
          )
          hdKey.derive('123')
        }).toThrow(/Path must start with "m" or "M"/)
      })
    })

    describe('- after wipePrivateData()', () => {
      it('should not have private data', () => {
        const hdkey = HDKey.fromMasterSeed(
          Buffer.from(keys.valid[6].seed, 'hex')
        ).wipePrivateData()
        expect(hdkey.privateKey).toEqual(null)
        expect(hdkey.privateExtendedKey).toEqual(null)
        expect(() => {
          hdkey.sign(Buffer.alloc(32))
        }).toThrow(/private key should be a Buffer/)
        const childKey = hdkey.derive('m/0')
        expect(childKey.publicExtendedKey).toEqual(keys.valid[7].public)
        expect(hdkey.privateKey).toEqual(null)
        expect(hdkey.privateExtendedKey).toEqual(null)
      })

      it('should have correct data', () => {
        // m/0/2147483647'/1/2147483646'/2
        const key =
          'xprvA2nrNbFZABcdryreWet9Ea4LvTJcGsqrMzxHx98MMrotbir7yrKCEXw7nadnHM8Dq38EGfSh6dqA9QWTyefMLEcBYJUuekgW4BYPJcr9E7j'
        const hdkey = HDKey.fromExtendedKey(key).wipePrivateData()
        expect(hdkey.versions.private).toEqual(0x0488ade4)
        expect(hdkey.versions.public).toEqual(0x0488b21e)
        expect(hdkey.depth).toEqual(5)
        expect(hdkey.parentFingerprint).toEqual(0x31a507b8)
        expect(hdkey.index).toEqual(2)

        expect(hdkey.chainCode.toString('hex')).toEqual(
          '9452b549be8cea3ecb7a84bec10dcfd94afe4d129ebfd3b3cb58eedf394ed271'
        )

        expect(hdkey.publicKey.toString('hex')).toEqual(
          '024d902e1a2fc7a8755ab5b694c575fce742c48d9ff192e63df5193e4c7afe1f9c'
        )

        expect(hdkey.identifier.toString('hex')).toEqual(
          '26132fdbe7bf89cbc64cf8dafa3f9f88b8666220'
        )

        expect(hdkey.pubKeyHash.toString('hex')).toEqual(
          '26132fdbe7bf89cbc64cf8dafa3f9f88b8666220'
        )
      })

      it('should be able to verify signatures', () => {
        const fullKey = HDKey.fromMasterSeed(
          Buffer.from(keys.valid[0].seed, 'hex')
        )
        // using JSON methods to clone before mutating
        const wipedKey = HDKey.fromJSON(fullKey.toJSON()).wipePrivateData()

        const hash = Buffer.alloc(32, 8)

        expect(wipedKey.verify(hash, fullKey.sign(hash))).toBeTruthy()
      })

      it('should not throw if called on hdkey without private data', () => {
        const hdkey = HDKey.fromExtendedKey(keys.valid[0].public)
        expect(() => hdkey.wipePrivateData()).not.toThrow()
        expect(hdkey.publicExtendedKey).toEqual(keys.valid[0].public)
      })
    })
  })
})


src/__test__/utils.spec.ts
import { hash160, serialize } from '../utils'

describe('utils', () => {
  it('hash160', () => {
    expect(
      hash160(
        Buffer.from(
          '02b4632d08485ff1df2db55b9dafd23347d1c47a457072a1e87be26896549a8737',
          'hex'
        )
      ).toString('hex')
    ).toEqual('93ce48570b55c42c2af816aeaba06cfee1224fae')
    expect(
      hash160(
        Buffer.from(
          '2c352c06030e090b212127390803312a1d22152c1d010d2c2811242c0d0f23372f21',
          'hex'
        )
      ).toString('hex')
    ).toEqual('2314e3e15dde32f6fbcbca85c2bf2b1e24abec74')
  })

  it('serialize/without parent', () => {
    expect(
      serialize(
        0x0488ade4,
        Buffer.from(
          '02b4632d08485ff1df2db55b9dafd23347d1c47a457072a1e87be26896549a8737',
          'hex'
        ),
        0,
        0,
        0x00000000,
        Buffer.from(
          '2c06030e090b212127390803312a1d22152c1d010d2c2811242c0d0f23372f21',
          'hex'
        )
      ).toString('hex')
    ).toEqual(
      '0488ade40000000000000000002c06030e090b212127390803312a1d22152c1d010d2c2811242c0d0f23372f2102b4632d08485ff1df2db55b9dafd23347d1c47a457072a1e87be26896549a8737'
    )
  })

  it('serialize/with parent', () => {
    expect(
      serialize(
        0x0488ade4,
        Buffer.from(
          '02b4632d08485ff1df2db55b9dafd23347d1c47a457072a1e87be26896549a8737',
          'hex'
        ),
        1,
        0,
        0x12345678,
        Buffer.from(
          '2c06030e090b212127390803312a1d22152c1d010d2c2811242c0d0f23372f21',
          'hex'
        )
      ).toString('hex')
    ).toEqual(
      '0488ade40112345678000000002c06030e090b212127390803312a1d22152c1d010d2c2811242c0d0f23372f2102b4632d08485ff1df2db55b9dafd23347d1c47a457072a1e87be26896549a8737'
    )
  })
})


src/utils.ts
import crypto from 'crypto'

const LEN = 78

export const hash160 = (buf: Buffer): Buffer => {
  const sha = crypto
    .createHash('sha256')
    .update(buf)
    .digest()
  return crypto
    .createHash('rmd160')
    .update(sha)
    .digest()
}

export const serialize = (
  version: number,
  key: Buffer,
  depth: number,
  index: number,
  parentFingerprint: number,
  chainCode: Buffer
) => {
  // => version(4) || depth(1) || fingerprint(4) || index(4) || chain(32) || key(33)

  const buffer = Buffer.allocUnsafe(LEN)

  buffer.writeUInt32BE(version, 0)
  buffer.writeUInt8(depth, 4)

  const fingerprint = depth ? parentFingerprint : 0x00000000

  buffer.writeUInt32BE(fingerprint, 5)
  buffer.writeUInt32BE(index, 9)

  chainCode.copy(buffer, 13)
  key.copy(buffer, 45)

  return buffer
}


src/index.ts
import assert from 'assert'
import cs from 'coinstring'
import crypto from 'crypto'
import secp256k1 from 'secp256k1'

import { hash160, serialize } from './utils'

const DEFAULT_VERSIONS: Versions = { private: 0x0488ade4, public: 0x0488b21e }
const DEFAULT_MASTER_SECRET = Buffer.from('Bitcoin seed', 'utf8')

class HDKey {
  public static HARDENED_OFFSET = 0x80000000

  public static fromMasterSeed(
    seed: Buffer,
    masterSecret?: string,
    versions = DEFAULT_VERSIONS
  ): HDKey {
    const secret = masterSecret
      ? Buffer.from(masterSecret, 'utf8')
      : DEFAULT_MASTER_SECRET

    const I = crypto
      .createHmac('sha512', secret)
      .update(seed)
      .digest()
    const IL = I.slice(0, 32)
    const IR = I.slice(32)

    const hdkey = new HDKey(versions)
    hdkey.chainCode = IR
    hdkey.privateKey = IL

    return hdkey
  }

  public static fromExtendedKey(
    base58Key: string,
    versions = DEFAULT_VERSIONS
  ): HDKey {
    // => version(4) || depth(1) || fingerprint(4) || index(4) || chain(32) || key(33)
    const hdkey = new HDKey(versions)

    const keyBuffer = cs.decode(base58Key)

    const version = keyBuffer.readUInt32BE(0)
    assert(
      version === versions.private || version === versions.public,
      'Version mismatch: does not match private or public'
    )

    hdkey._depth = keyBuffer.readUInt8(4)
    hdkey._parentFingerprint = keyBuffer.readUInt32BE(5)
    hdkey._index = keyBuffer.readUInt32BE(9)
    hdkey._chainCode = keyBuffer.slice(13, 45)

    const key = keyBuffer.slice(45)
    if (key.readUInt8(0) === 0) {
      // private
      assert(
        version === versions.private,
        'Version mismatch: version does not match private'
      )
      hdkey.privateKey = key.slice(1) // cut off first 0x0 byte
    } else {
      assert(
        version === versions.public,
        'Version mismatch: version does not match public'
      )
      hdkey.publicKey = key
    }

    return hdkey
  }

  public static fromJSON(obj: HDKeyJSON): HDKey {
    return HDKey.fromExtendedKey(obj.xpriv)
  }

  private _versions: Versions
  private _depth: number
  private _index: number
  private _privateKey: Buffer
  private _publicKey: Buffer
  private _chainCode: Buffer
  private _fingerprint: number
  private _parentFingerprint: number
  private _identifier: Buffer

  constructor(versions = DEFAULT_VERSIONS) {
    this._versions = versions
    this._depth = 0
    this._index = 0
    this._privateKey = null
    this._publicKey = null
    this._chainCode = null
    this._fingerprint = 0
    this._parentFingerprint = 0
  }

  get versions(): Versions {
    return this._versions
  }

  get depth(): number {
    return this._depth
  }

  get index(): number {
    return this._index
  }

  get parentFingerprint(): number {
    return this._parentFingerprint
  }

  get fingerprint(): number {
    return this._fingerprint
  }

  get identifier(): Buffer {
    return this._identifier
  }

  get pubKeyHash(): Buffer {
    return this._identifier
  }

  get privateKey(): Buffer {
    return this._privateKey
  }

  set privateKey(key: Buffer) {
    assert.equal(key.length, 32, 'Private key must be 32 bytes.')
    assert(secp256k1.privateKeyVerify(key) === true, 'Invalid private key')

    this._privateKey = key
    this._publicKey = secp256k1.publicKeyCreate(key, true)
    this._identifier = hash160(this._publicKey)
    this._fingerprint = this._identifier.slice(0, 4).readUInt32BE(0)
  }

  get publicKey(): Buffer {
    return this._publicKey
  }

  set publicKey(key: Buffer) {
    assert(
      key.length === 33 || key.length === 65,
      'Public key must be 33 or 65 bytes.'
    )
    assert(secp256k1.publicKeyVerify(key) === true, 'Invalid public key')

    this._publicKey = secp256k1.publicKeyConvert(key, true) // force compressed point
    this._identifier = hash160(this.publicKey)
    this._fingerprint = this._identifier.slice(0, 4).readUInt32BE(0)
    this._privateKey = null
  }

  get chainCode(): Buffer {
    return this._chainCode
  }

  set chainCode(code: Buffer) {
    this._chainCode = code
  }

  get privateExtendedKey(): string {
    if (this._privateKey) {
      return cs.encode(
        serialize(
          this._versions.private,
          Buffer.concat([Buffer.alloc(1, 0), this._privateKey]),
          this._depth,
          this._index,
          this._parentFingerprint,
          this._chainCode
        )
      )
    }
    return null
  }

  get publicExtendedKey(): string {
    return cs.encode(
      serialize(
        this._versions.public,
        this._publicKey,
        this._depth,
        this._index,
        this._parentFingerprint,
        this._chainCode
      )
    )
  }

  public derive(path: string): HDKey {
    if (path.toLowerCase() === 'm' || path.toLowerCase() === "m'") {
      return this
    }

    let newHDKey: HDKey = HDKey.fromExtendedKey(
      this.privateExtendedKey || this.publicExtendedKey,
      this.versions
    )

    path.split('/').forEach((c, index) => {
      if (index === 0) {
        assert(/^[mM]{1}/.test(c), 'Path must start with "m" or "M"')
        return
      }

      const hardened = c.length > 1 && c.includes("'")

      let childIndex = parseInt(c, 10)
      assert(childIndex < HDKey.HARDENED_OFFSET, 'Invalid index')

      if (hardened) {
        childIndex += HDKey.HARDENED_OFFSET
      }

      newHDKey = newHDKey.deriveChild(childIndex)
    })

    return newHDKey
  }

  public deriveChild(index: number): HDKey {
    const isHardened = index >= HDKey.HARDENED_OFFSET
    const indexBuffer = Buffer.allocUnsafe(4)
    indexBuffer.writeUInt32BE(index, 0)

    let data: Buffer

    if (isHardened) {
      // Hardened child
      assert(this._privateKey, 'Could not derive hardened child key')

      const zb = Buffer.alloc(1, 0)
      const pk = Buffer.concat([zb, this._privateKey])

      // data = 0x00 || ser256(kpar) || ser32(index)
      data = Buffer.concat([pk, indexBuffer])
    } else {
      // Normal child
      // data = serP(point(kpar)) || ser32(index)
      //      = serP(Kpar) || ser32(index)
      data = Buffer.concat([this.publicKey, indexBuffer])
    }

    const I = crypto
      .createHmac('sha512', this._chainCode)
      .update(data)
      .digest()
    const IL = I.slice(0, 32)
    const IR = I.slice(32)

    const newHDKey = new HDKey(this.versions)

    // Private parent key -> private child key
    if (this._privateKey) {
      // ki = parse256(IL) + kpar (mod n)
      try {
        newHDKey.privateKey = secp256k1.privateKeyTweakAdd(this.privateKey, IL)
        // throw if IL >= n || (privateKey + IL) === 0
      } catch (err) {
        // In case parse256(IL) >= n or ki == 0, one should proceed with the next value for i
        return this.deriveChild(index + 1)
      }
      // Public parent key -> public child key
    } else {
      // Ki = point(parse256(IL)) + Kpar
      //    = G*IL + Kpar
      try {
        newHDKey.publicKey = secp256k1.publicKeyTweakAdd(
          this.publicKey,
          IL,
          true
        )
        // throw if IL >= n || (g**IL + publicKey) is infinity
      } catch (err) {
        // In case parse256(IL) >= n or Ki is the point at infinity, one should proceed with the next value for i
        return this.deriveChild(index + 1)
      }
    }

    newHDKey._chainCode = IR
    newHDKey._depth = this._depth + 1
    newHDKey._parentFingerprint = this._fingerprint
    newHDKey._index = index

    return newHDKey
  }

  public sign(hash: Buffer): Buffer {
    return secp256k1.sign(hash, this.privateKey).signature
  }

  public verify(hash: Buffer, signature: Buffer): boolean {
    return secp256k1.verify(hash, signature, this._publicKey)
  }

  public wipePrivateData(): HDKey {
    if (this._privateKey) {
      crypto.randomBytes(this._privateKey.length).copy(this._privateKey)
    }
    this._privateKey = null
    return this
  }

  public toJSON(): HDKeyJSON {
    return {
      xpriv: this.privateExtendedKey,
      xpub: this.publicExtendedKey
    }
  }
}

export interface Versions {
  private: number
  public: number
}

export interface HDKeyJSON {
  xpriv: string
  xpub: string
}

export default HDKey
